### PR TITLE
loading spinner centering

### LINKED ISSUE #280: UX Bug — Explore load page spinner offcenter
**Describe the bug**
The spinner for the loading pages in explore are off. 
Screen Shot 2020-10-31 at 1 12 02 AM

**To Reproduce**
1. Go to explore
2. Loading screen

**Expected behavior**
These apply to the loading pages for CourseExplorer, SubjectExplorer, and CoursePage
1. The page content should fill 100% page height such that the footer is below the viewport
2. The loading spinner should be in the center of the page

### PR DESCRIPTION
**Pull request by kylezehaoqin**
**Issue**
Closes #280 
Loading spinner is not centered and footer wasn't remained at the page bottom

**Database Changes/Migrations**
N/A

**Test Modifications**

N/A

**Test Procedure**

run `docker volume rm yacsn_postgres_data` to clear database, or manually stop yacs_api in docker dashboard.
Navigate to explore page
Observe loading spinner and footer position 

**Photos**
Before
image

After
image

**Additional Info**

N/A

### CODE REVIEW

**Comment by jshom** [comment]
Seeing duplicate styles, maybe we make a centerredSpinner component? What do you think?

**Comment by kylezehaoqin** [comment]
That is a good point, I'll work on it

**Review by jshom — CHANGES_REQUESTED** [review]
Nice! few small things:

Take a look at: https://66.228.32.91/?semester=SPRING%202021, looks like on load it snaps. My guess is it has to do with the preset height.

Lets parametrize the component with 2 things:
1. The `height: {X}vh` param
2. The loading message. So we can change "Loading Departments" and "Loading Courses" with just a change of params
3. The font-size (take a look at loading here: yacs.cs.rpi.edu/, appears the font size is lower, my guess 1em)

**Comment by kylezehaoqin** [comment]
> Nice! few small things:
> 
> Take a look at: https://66.228.32.91/?semester=SPRING%202021, looks like on load it snaps. My guess is it has to do with the preset height.
> 
> Lets parametrize the component with 3 things:
> 
> 1. The `height: {X}vh` param
> 2. The loading message. So we can change "Loading Departments" and "Loading Courses" with just a change of params
> 3. The font-size (take a look at loading here: https://yacs.cs.rpi.edu/, appears the font size is lower, my guess 1em)

Now the component should be all set. I also passed in a top padding parameter so the loading message won't leave a big chunk of white space for the rest of the viewport above footer.

File: src/web/src/pages/CourseExplorer.vue [hunk 1]
```diff
@@ -56,9 +56,13 @@
       </b-col>
     </b-row>
   </b-container>
-  <div v-else>
-    <b-spinner></b-spinner>
-    <strong class="m-2">Loading courses...</strong>
+  <div id="spinnerWrapper" v-else>
```
Comment (IkeLyons) [inline src/web/src/pages/CourseExplorer.vue:59]: Do you need this div here? Seems like you might be able to just do `<CenterSpinner v-else`. If the div is needed though you should still remove the `spinnerWrapper` ID as its not being used. 

This comment also applies to the CoursePage, SubjectExplorer, and CourseScheduler

File: src/web/src/components/CenterSpinner.vue [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  },
+  computed: {
+    renderStyle() {
+      return "padding-top: " + this.topSpacing + "vh;" + 
+             "font-size: " + this.fontSize + "em;" +
+             "height: " + this.height + "vh;" +
+             "text-align: center;";
+    }
+  },
+  async created() {
```
Comment (IkeLyons) [inline src/web/src/components/CenterSpinner.vue:25]: This created should be deleted as it is just logging.

**Review by IkeLyons — CHANGES_REQUESTED** [review]
Couple changes but otherwise looks good!

File: src/web/src/pages/CourseExplorer.vue (hunk 1 above)
Comment (kylezehaoqin) [reply src/web/src/pages/CourseExplorer.vue:59]: I forgot that vue3 doesn't require you to have only one parental directory under template, updated.

File: src/web/src/pages/CoursePage.vue [hunk 3]
```diff
@@ -25,18 +25,25 @@
...
   </div>
-  <div v-else>
-    <b-spinner></b-spinner>
-    <strong class="m-2">Loading Course...</strong>
-  </div>
+  <CenterSpinner 
+    v-else
+    :height = "80" 
+    :fontSize = "1.4"
+    :loadingMessage = "Courses"
```
Comment (IkeLyons) [inline src/web/src/pages/CoursePage.vue:32]: This colon needs to be removed, its interpreting `"Courses"` as js code so the message isn't being displayed.

**Review by IkeLyons — CHANGES_REQUESTED** [review]
Sorry two last things. One I commented on below, and the other is that I'm noticing the spinner on the subject explorer doesn't seem to be vertically centered? The other pages are good, but subject explorer's is at the top of the page. 

Once these are fixed should be good to merge!

File: src/web/src/pages/CoursePage.vue (hunk 3 above)
Comment (kylezehaoqin) [reply src/web/src/pages/CoursePage.vue]: You're good I was being ridiculously careless. The spinner for subject explorer should also be fixed. Thanks for pointing these out!

**Comment by marchdan** [comment]
@kylezehaoqin Please resolve conflicts so we can review :)

**Review by YuetingLiao — COMMENTED** [review]
Nice done! So all page loadings will be in the center of the page now. It could be merged in after fix minor issues and refactor some duplicate code mentioned by Jacob.

**Review by IkeLyons — APPROVED** [review]

**Review by 123joshuawu — APPROVED** [review]
looks good to me!

one note, instead of providing the styles as a string, you can also provide them as an object: https://vuejs.org/v2/guide/class-and-style.html#Object-Syntax-1

so instead of this 
```javascript
return "padding-top: " + this.topSpacing + "vh;" + 
           "font-size: " ...
```

it would look like this
```javascript
return {
  "padding-top": this.topSpacing + "vh",
  "font-size": this.fontSize + "em",
  ...
}
```

**Comment by marchdan** [comment]
@jshom You need to close your change requests before this can be merged

**Review by jshom — APPROVED** [review]

**Comment by jshom** [comment]
Agree with @123joshuawu comment, lets get that in for code quality before merging @kylezehaoqin

**Comment by jshom** [comment]
Awesome stuff on this!

**Pull request merged**